conjure_oxide/src/unstable/solver_interface/states.rs

```diff
@@ -0,0 +1,42 @@
+//! States of a [`Solver`].
+use std::fmt::Display;
+
+use thiserror::Error;
+
+use super::private::Internal;
+use super::private::Sealed;
+use super::stats::*;
+use super::Solver;
+use super::SolverError;
+
+pub trait SolverState: Sealed {}
+
+impl Sealed for Init {}
+impl Sealed for ModelLoaded {}
+impl Sealed for ExecutionSuccess {}
+impl Sealed for ExecutionFailure {}
+
+impl SolverState for Init {}
+impl SolverState for ModelLoaded {}
+impl SolverState for ExecutionSuccess {}
+impl SolverState for ExecutionFailure {}
+
+pub struct Init;
+pub struct ModelLoaded;
+
+/// The state returned by [`Solver`] if solving has been successful.
+pub struct ExecutionSuccess {
+    /// Execution statistics.
+    pub stats: Option<Box<dyn Stats>>,
+
+    /// Cannot construct this from outside this module.
+    pub _sealed: Internal,
+}
+
+/// The state returned by [`Solver`] if solving has not been successful.
```

Comment:
What does "not successful" mean? Is it some kind of solver error, or there just weren't any solutions? If it's "no solutions", did we just run out of time (if there was a time limit)? I'd be clearer which is going on

Reply:
Maybe I could have type states for SolverFinish and SolverCrash, and then we can subdivide SolverFinish into an arbitrary number of  subclasses such as sat, unsat, timeout, ...

Reply:
I'd keep solver errors seperate, because I suspect most people will want to `unwrap` those (because solver error generally means something horrible has happened, and it's probably best to just abort the program altogether).

With regards timeout vs solutions, there isn't a perfect answer. I'd be tempted to just always return a list of solutions (which might be empty, or have 1, or have many, depending on what was asked for). While a list of length 1 might be annoying for people who only want one solution, it's also a bit annoying to split 1 solutions vs many (unless there is going to be different search methods to start a search for 1 solution vs many?) and then also have SolverFinished split into complete or incomplete (incomplete can be because of timeout, or some kind of "search/node limit").

Reply:
I agree with all this, thank you for the feedback :)

We are getting solutions out by callback, and give the user custom logic to handle/ not handle them / abort search early - therefore I don't think we can have a list of them in a clean way, and the user should do this themselves if they want it.
 
I think we decided not to give a list for memory concerns as a model can have a large number of solutions - a good user coded callback could just drop solutions deemed to be identical or useless from the list, but we can't do that in generality (at this level of the program).


However I can definitely keep track of if any solutions have been found, and say "you got solutions" vs "you got no solutions". 

Incomplete and complete is definitely a  good taxonomy - would "user aborted the search" be counted as complete?